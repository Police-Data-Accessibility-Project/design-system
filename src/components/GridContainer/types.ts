export interface PdapGridContainerProps {
	className?: string;
	columns?: 1 | 2 | 3 | 'auto';
	component?: string;
	rows?: number | 'auto';
	templateColumns?: string;
	templateRows?: string;
}
