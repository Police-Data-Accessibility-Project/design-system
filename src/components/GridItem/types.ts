export interface PdapGridItemProps {
	className?: string;
	component?: string;
	spanColumn?: 1 | 2 | 3;
	spanRow?: number;
}
