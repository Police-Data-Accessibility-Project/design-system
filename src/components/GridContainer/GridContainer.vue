<template>
	<component :is="component" :class="classes" :style="styles">
		<slot />
	</component>
</template>

<script setup lang="ts">
import { reactive } from 'vue';

import { PdapGridContainerProps } from './types';

const props = withDefaults(defineProps<PdapGridContainerProps>(), {
	columns: 'auto',
	component: 'div',
	rows: 'auto',
});

// CSS class map
const classes = reactive({
	'pdap-grid-container': true,
	[`pdap-grid-container-column-${props.columns}`]: props.columns !== 'auto',
	[String(props.className)]: Boolean(props.className),
});
// CSS styles map
const renderTemplateRowsInline =
	props.templateRows || typeof props.rows === 'number';
const gridTemplateRows =
	props.templateRows ?? `repeat(${props.rows}, minmax(0, 1fr))`;
const styles = reactive({
	...(renderTemplateRowsInline && {
		gridTemplateRows,
	}),
	...(props.templateColumns && { gridTemplateColumns: props.templateColumns }),
});
</script>

<style scoped>
.pdap-grid-container {
	@apply grid grid-cols-[auto] grid-rows-[auto] h-full gap-4 p-8 w-full;
	@apply md:gap-8;
}

.pdap-grid-container-column-1 {
	@apply grid-cols-1;
}

.pdap-grid-container-column-2 {
	@apply items-start grid-cols-1;
	@apply md:grid-cols-2;
}

.pdap-grid-container-column-3 {
	@apply grid-cols-1;
	@apply md:grid-cols-2 lg:grid-cols-3;
}
</style>
