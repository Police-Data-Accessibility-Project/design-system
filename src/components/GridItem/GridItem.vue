<template>
	<component :is="component" :class="classes" :style="styles">
		<slot />
	</component>
</template>

<script setup lang="ts">
import { reactive } from 'vue';

import { PdapGridItemProps } from './types';

const props = withDefaults(defineProps<PdapGridItemProps>(), {
	component: 'div',
	spanColumn: 1,
	spanRow: 1,
});

// CSS class map
const classes = reactive({
	'pdap-grid-item': true,
	[`pdap-grid-item-span-column-${props.spanColumn}`]: props.spanColumn > 1,
	[String(props.className)]: Boolean(props.className),
});
// CSS styles
const styles = reactive({
	// Only add inline property if it is passed and exceeds the default
	...(props.spanRow > 1 && {
		gridRow: `span ${props.spanRow} / span ${props.spanRow}`,
	}),
});
</script>

<style scoped>
.pdap-grid-item {
	@apply col-span-1 row-span-1;
}
.pdap-grid-item-span-column-2 {
	@apply md:col-span-2;
}

.pdap-grid-item-span-column-3 {
	@apply col-span-1;
	@apply md:col-span-2 lg:col-span-3;
}
</style>
