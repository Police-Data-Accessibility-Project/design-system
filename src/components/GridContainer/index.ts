import GridContainer from './GridContainer.vue';

export { GridContainer };
