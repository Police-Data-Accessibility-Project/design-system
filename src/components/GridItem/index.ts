import GridItem from './GridItem.vue';

export { GridItem };
