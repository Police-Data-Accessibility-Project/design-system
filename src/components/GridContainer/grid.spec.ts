/* eslint-disable vue/one-component-per-file */
// Component
import GridContainer from './GridContainer.vue';
import GridItem, { PdapGridItemProps } from '../GridItem/GridItem.vue';

// Utils
import { defineComponent } from 'vue';
import { mount } from '@vue/test-utils';
import { describe, expect, test, vi } from 'vitest';

type TestSlotProps<T extends PdapGridItemProps = PdapGridItemProps> = T;

const data = [
	{ component: 'li' },
	{ component: 'img', src: 'https://mock.test.com' },
	{ component: 'card', spanColumn: 3, spanRow: 2 },
];
const template =
	'<GridItem v-for="item in data" :key="`item.component + ${item.spanRow * Math.random()}`" :component="item.component" :src="item.src ?? undefined" :spanRow="item.spanRow"><slot /></GridItem>';

const expectComponents = new Set(['li', 'img', 'card']);

const MultipleItems = defineComponent({
	components: {
		GridItem,
	},
	props: {
		data: {
			type: Array<TestSlotProps>,
			default: data,
		},
	},
	template,
});

const SingleItem = defineComponent({
	components: {
		GridItem,
	},
	props: {
		data: {
			type: Array<TestSlotProps>,
			default: data,
		},
	},
	template: '<GridItem :spanRow="3"><slot /></GridItem>',
});

const getComputedStyle = vi.fn((el) => el.getComputedStyle());
getComputedStyle.mockReturnValue({
	gridTemplateRows: 'repeat(3, minmax(20px, 1fr))',
	gridTemplateColumns: 'repeat(2, auto)',
	gridRow: 'span 2 / span 2',
});
vi.stubGlobal('getComputedStyle', getComputedStyle);

// Test
describe('Renders container component', () => {
	// Render
	test('Renders a container', () => {
		const wrapper = mount(GridContainer, {
			slots: {
				default: MultipleItems,
			},
		});

		const container = wrapper.find('.pdap-grid-container');
		const items = container.findAll('.pdap-grid-item');

		// Container
		expect(container.exists()).toBe(true);
		expect(wrapper.classes()).toContain('pdap-grid-container');

		// Renders all items passed
		expect(items.length).toBe(3);

		// Renders fall-through prop conditionally for img element
		expect(container.find('[src="https://mock.test.com"]').exists()).toBe(true);

		// Renders each type of component passed
		expect(
			items.every((item) =>
				item.getRootNodes().some((node) => expectComponents.has(node.localName))
			)
		).toBe(true);

		// Snapshot
		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Renders a container with template props passed', () => {
		const wrapper = mount(GridContainer, {
			slots: {
				default: MultipleItems,
			},
			props: {
				templateColumns: 'repeat(2, auto)',
				templateRows: 'repeat(3, minmax(20px, 1fr))',
			},
		});

		expect(window.getComputedStyle(wrapper.vm.$el).gridTemplateColumns).toBe(
			'repeat(2, auto)'
		);
		expect(window.getComputedStyle(wrapper.vm.$el).gridTemplateRows).toBe(
			'repeat(3, minmax(20px, 1fr))'
		);
	});

	test('Renders a container with columns and rows props passed', () => {
		const wrapper = mount(GridContainer, {
			slots: {
				default: MultipleItems,
			},
			props: {
				rows: 2,
				columns: 3,
			},
		});

		expect(wrapper.vm.$props.rows).toBe(2);
		expect(wrapper.vm.$props.columns).toBe(3);
	});

	test('Renders a grid item with custom row span prop', () => {
		const wrapper = mount(GridContainer, {
			slots: {
				default: SingleItem,
			},
		});

		expect(
			window.getComputedStyle(wrapper.find('.pdap-grid-item').element).gridRow
		).toBe('span 2 / span 2');
	});
});
