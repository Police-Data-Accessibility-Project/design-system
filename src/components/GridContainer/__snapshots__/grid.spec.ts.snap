// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders container component > Renders a container 1`] = `
<div class="pdap-grid-container">
  <li class="pdap-grid-item"></li>
  <img class="pdap-grid-item" src="https://mock.test.com">
  <card class="pdap-grid-item" style="grid-row: span 2 / span 2;"></card>
</div>
`;
